### DISCUSSION TITLE
Does RAND_bytes() use /dev/random, or /dev/urandom() on Linux ?

### DISCUSSION THREAD

**Discussion by vk18anubhav** [openssl/openssl#25951] [Q&A]
Does RAND_bytes() use /dev/random, or /dev/urandom() on Linux ? Is this in some way configureable ?

Is the behaviour of RAND_bytes() same for FIPS and non-FIPS mode ?

I am using OpenSSL 3.0.15

**Comment by mattcaswell** [comment]
> Does RAND_bytes() use /dev/random, or /dev/urandom() on Linux ? Is this in some way configureable ?

You can configure the seed sources at build time of OpenSSL.

By default it will use the "os-specific" seed source. On Linux this means it will attempt to seed using the "getrandom" syscall in preference (if the kernel supports it). Otherwise it will fall back to some /dev/*random device the order of preference being: "/dev/urandom", "/dev/random", "/dev/hwrng", "/dev/srandom"

See:

https://github.com/openssl/openssl/blob/master/INSTALL.md#seeding-the-random-generator

**Comment by mattcaswell** [comment]
> Is the behaviour of RAND_bytes() same for FIPS and non-FIPS mode ?

The FIPS provider does not contain a seed source, so ultimately it will fall back to the same seed source as with non-FIPS. However the seed source is only a seed. The DRBG in the FIPS provider will use that seed to then generate that actual final random number that `RAND_bytes()` returns. The DRBG in the FIPS provider is validated to FIPS standards and therefore maybe different to the one you get in the default provider.

**Reply by paulidale** [reply]
RAND_bytes() will always use a DRBG even out of FIPS mode.  The raw bytes used for seeding are never directly returned.

**Reply by bukka** [reply]
Just for visibility adding that it's possible to not use DRBG with my provider https://github.com/bukka/osrand-provider if OS-RAND random is selected.

**Comment by vk18anubhav** [comment]
Hi Everyone,

Thanks for prompt response on last query.

 What “os-specific” interfaces RAND_bytes() uses on other platforms like HP-UX, Solaris, AIX etc ?

**Reply by paulidale** [reply]
Look in [providers/implementations/rands/seeding/rand_unix.c](https://github.com/openssl/openssl/blob/master/providers/implementations/rands/seeding/rand_unix.c)
